iOS Client/Store/MBArticlesStore.swift

```diff
@@ -17,184 +17,172 @@ class MBArticlesStore: NSObject {
         super.init()
     }
     
-    /***** Read Data from Core Data (Only Invoke These From The Main Thread) *****/
-    func getAuthors(persistentContainer: NSPersistentContainer) -> [MBAuthor] {
+    /***** Read from Core Data *****/
+    func getAuthors(managedObjectContext: NSManagedObjectContext) -> [MBAuthor] {
         let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: MBAuthor.entityName)
-        return performFetch(managedContext: persistentContainer.viewContext, fetchRequest: fetchRequest) as? [MBAuthor] ?? []
+        return performFetch(managedContext: managedObjectContext, fetchRequest: fetchRequest) as? [MBAuthor] ?? []
     }
     
-    func getCategories(persistentContainer: NSPersistentContainer) -> [MBCategory] {
+    func getCategories(managedObjectContext: NSManagedObjectContext) -> [MBCategory] {
         let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: MBCategory.entityName)
-        return performFetch(managedContext: persistentContainer.viewContext, fetchRequest: fetchRequest) as? [MBCategory] ?? []
+        return performFetch(managedContext: managedObjectContext, fetchRequest: fetchRequest) as? [MBCategory] ?? []
     }
     
-    func getArticles(persistentContainer: NSPersistentContainer) -> [MBArticle] {
+    func getArticles(managedObjectContext: NSManagedObjectContext) -> [MBArticle] {
         let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: MBArticle.entityName)
-        return performFetch(managedContext: persistentContainer.viewContext, fetchRequest: fetchRequest) as? [MBArticle] ?? []
+        return performFetch(managedContext: managedObjectContext, fetchRequest: fetchRequest) as? [MBArticle] ?? []
     }
     
     private func performFetch(managedContext: NSManagedObjectContext, fetchRequest: NSFetchRequest<NSManagedObject>) -> [NSManagedObject] {
-        do {
-            return try managedContext.fetch(fetchRequest)
-        } catch let error as NSError {
-            print("Could not fetch. \(error), \(error.userInfo)")
-            return []
+        var retVal: [NSManagedObject] = []
+        
+        managedContext.performAndWait {
+            do {
+                retVal = try managedContext.fetch(fetchRequest)
+            } catch let error as NSError {
+                print("Could not fetch. \(error), \(error.userInfo)")
+                retVal = []
+            }
         }
+        
+        return retVal
     }
     
-    /***** Download Data from Network (MAY BE INVOKED FROM ANY THREAD) *****/
-    func syncAllData(persistentContainer: NSPersistentContainer, completion: @escaping (Bool?, Error?) -> Void) {
```

Comment:
I think if we introduce PromiseKit or PinkyPromise then this method will be much more manageable. Right now it is a little too much nesting for me